lib/Transforms/Utils/Local.cpp

```diff
@@ -332,9 +335,19 @@ bool llvm::isInstructionTriviallyDead(Instruction *I,
     if (Constant *C = dyn_cast<Constant>(CI->getArgOperand(0)))
       return C->isNullValue() || isa<UndefValue>(C);
 
-  // HLSL change - don't force unused convergenet markers to stay
-  if (CallInst *CI = dyn_cast<CallInst>(I))
+  // HLSL Change - don't force unused convergenet markers to stay, 
+  // remove bad OutputCompletes
+  if (CallInst *CI = dyn_cast<CallInst>(I)) {  
     if (hlsl::dxilutil::IsConvergentMarker(CI)) return true;
+    if (hlsl::OP::IsDxilOpFuncCallInst(I, hlsl::OP::OpCode::OutputComplete)) {
```

Comment:
I wonder if we should instead call a utility function (perhaps defined in `dxilutil`) which returns true if the function call is an intermediate or DXIL op that has no side effects, for functions not marked ReadNone or ReadOnly.  This would incorporate both checks, but could also be expanded to handle more, such as wave ops.

Reply:
This is for our operations that otherwise appear to have side effects (aren't marked as (ReadOnly or ReadNone) because we want standard optimizations to treat them as if they have side effects.  But if the value is unused, or in this case, the handle is null, the operation is actually a no-op and does not have side effects.

Reply:
And `mayHaveSideEffects()` is an inexpensive inline operation on `Instruction` that we wouldn't want to pollute with DXIL operation checks and such.